lmdeploy/utils.py

```diff
@@ -516,22 +517,28 @@ def __init__(
             metadata: Pre-computed metadata (for reconstruction)
         """
         if named_tensors is not None:
-            metadata_li = [None] * len(named_tensors)
-            flattened_tensor_li = [None] * len(named_tensors)
-            current_idx = 0
-            for idx, (name, tensor) in enumerate(named_tensors):
-                flattened_tensor_li[idx] = tensor.flatten()
-                numel = tensor.numel()
-                metadata_li[idx] = FlattenedTensorMetadata(name=name,
-                                                           shape=tensor.shape,
-                                                           dtype=tensor.dtype,
-                                                           start_idx=current_idx,
-                                                           end_idx=current_idx + numel,
-                                                           numel=numel)
-                current_idx += numel
-
-            self.metadata = metadata_li
-            self.flattened_tensor = torch.cat(flattened_tensor_li, dim=0)
+            num_tensors = len(named_tensors)
+            self.metadata = [None] * num_tensors
+            self.flattened_tensor = [None] * num_tensors
+            if num_tensors > 0:
+                if num_tensors > 1:
+                    dtypes = [t.dtype for _, t in named_tensors]
+                    if not all([d == dtypes[0] for d in dtypes[1:]]):
+                        raise ValueError(f'All tensors should have same dtype, but given {dtypes}')
```

Comment:
Is it correct to ensure all tensors are of the same dtype? Take `Qwen3-32B-FP8` as an example,  the dtype of `model.embed_tokens.weight` and `model.layers.0.input_layernorm.weight` are bfloat16, but the dtype of `model.layers.6.self_attn.q_proj.weight` and `model.layers.0.mlp.down_proj.weight` are float8_e4m3fn

Reply:
If the tensors in `self.flattened_tensor` have different data types, `torch.cat()` will attempt automatic type promotion according to PyTorch's type promotion rules, which might bring exception.
```python
self.flattened_tensor = torch.cat(self.flattened_tensor, dim=0)
```